src/Generics/Deriving/TH/Internal.hs

```diff
@@ -145,52 +145,65 @@ makeFunKind = makeFunType
 makeFunKind argKinds resKind = foldr' ArrowK resKind argKinds
 #endif
 
--- | Detect if a Name occurs as an argument to some type family. This makes an
--- effort to exclude /oversaturated/ arguments to type families. For instance,
--- if one declared the following type family:
---
--- @
--- type family F a :: Type -> Type
--- @
---
--- Then in the type @F a b@, we would consider @a@ to be an argument to @F@,
--- but not @b@.
-isInTypeFamilyApp :: Name -> Type -> [Type] -> Q Bool
-isInTypeFamilyApp name tyFun tyArgs =
-  case tyFun of
-    ConT tcName -> go tcName
-    _           -> return False
+-- | Remove any outer `SigT` and `ParensT` constructors, and turn
+-- an outermost `InfixT` constructor into plain applications.
+dustOff :: Type -> Type
+dustOff (SigT ty _) = dustOff ty
+#if MIN_VERSION_template_haskell(2,11,0)
+dustOff (ParensT ty) = dustOff ty
+dustOff (InfixT ty1 n ty2) = ConT n `AppT` ty1 `AppT` ty2
+#endif
+dustOff ty = ty
+
+-- | Checks whether a type is an unsaturated type family
+-- application.
+isUnsaturatedType :: Type -> Q Bool
+isUnsaturatedType = go 0 . dustOff
   where
-    go :: Name -> Q Bool
-    go tcName = do
+    -- Expects its argument to be dusted
+    go :: Int -> Type -> Q Bool
+    go d t = case t of
+      ConT tcName -> check d tcName
+      AppT f _ -> go (d + 1) (dustOff f)
+      _ -> return False
+
+    check :: Int -> Name -> Q Bool
+    check d tcName = do
+      mbinders <- getTypeFamilyBinders tcName
+      return $ case mbinders of
+        Just bndrs -> length bndrs > d
+        Nothing -> False
+
+-- | Given a name, check if that name is a type family. If
+-- so, return a list of its binders.
+#if MIN_VERSION_template_haskell(2,17,0)
+getTypeFamilyBinders :: Name -> Q (Maybe [TyVarBndr ()])
```

Comment:
You can avoid guarding this with CPP by using `TyVarBndr_ ()` instead, where [`TyVarBndr_`](https://hackage.haskell.org/package/th-abstraction-0.4.3.0/docs/Language-Haskell-TH-Datatype-TyVarBndr.html#t:TyVarBndr_) comes from `th-abstraction`.